zingo-sync/src/scan.rs

```diff
@@ -524,3 +545,143 @@ fn collect_nullifiers(
+            if !wallet_block.txids().contains(&transaction.txid()) {
+                panic!("txid is not found in the wallet block at the transaction height!");
+            }
+        } else {
+            panic!("wallet block at transaction height not found!");
+        }
+
+        let wallet_transaction = scan_transaction(
+            parameters,
+            scanning_keys,
+            transaction,
+            block_height,
+            &decrypted_note_data,
+        )
+        .unwrap();
+        wallet_transactions.insert(txid, wallet_transaction);
+    }
+
+    Ok(wallet_transactions)
+}
+
+fn scan_transaction<P>(
+    parameters: &P,
+    scanning_keys: &ScanningKeys,
+    transaction: Transaction,
+    block_height: BlockHeight,
+    decrypted_note_data: &DecryptedNoteData,
+) -> Result<WalletTransaction, ()>
+where
+    P: Parameters,
+{
+    // TODO: price?
+    let zip212_enforcement = zip212_enforcement(parameters, block_height);
+    let mut sapling_notes: Vec<SaplingNote> = Vec::new();
+    let mut orchard_notes: Vec<OrchardNote> = Vec::new();
+
+    if let Some(bundle) = transaction.sapling_bundle() {
+        let sapling_keys: Vec<sapling_crypto::keys::PreparedIncomingViewingKey> = scanning_keys
+            .sapling()
+            .iter()
+            .map(|(_, key)| key.prepare())
+            .collect();
+        let sapling_outputs: Vec<(SaplingDomain, OutputDescription<GrothProofBytes>)> = bundle
+            .shielded_outputs()
+            .iter()
+            .map(|output| (SaplingDomain::new(zip212_enforcement), output.clone()))
+            .collect();
+
+        scan_notes::<SaplingDomain, OutputDescription<GrothProofBytes>, SaplingNote>(
+            &mut sapling_notes,
+            transaction.txid(),
+            &sapling_keys,
+            &sapling_outputs,
+            &decrypted_note_data.sapling_nullifiers_and_positions,
+        )
+        .unwrap();
+    }
+
+    if let Some(bundle) = transaction.orchard_bundle() {
+        let orchard_keys: Vec<orchard::keys::PreparedIncomingViewingKey> = scanning_keys
+            .orchard()
+            .iter()
+            .map(|(_, key)| key.prepare())
+            .collect();
+        let orchard_actions: Vec<(OrchardDomain, Action<Signature<SpendAuth>>)> = bundle
+            .actions()
+            .iter()
+            .map(|action| (OrchardDomain::for_action(action), action.clone()))
+            .collect();
+
+        scan_notes::<OrchardDomain, Action<Signature<SpendAuth>>, OrchardNote>(
+            &mut orchard_notes,
+            transaction.txid(),
+            &orchard_keys,
+            &orchard_actions,
+            &decrypted_note_data.orchard_nullifiers_and_positions,
+        )
+        .unwrap();
+    }
+
+    Ok(WalletTransaction::from_parts(
+        transaction.txid(),
+        block_height,
+        sapling_notes,
+        orchard_notes,
+    ))
+}
+
+fn scan_notes<D, Op, N>(
+    wallet_notes: &mut Vec<N::WalletNote>,
```

Comment:
why do we need to pass a mutable vector here? I see that the input is empty. instead of the extra parameter just initialize an empty vector in the function

Reply:
this way is the simplest and cleanest because the vec is initialised in the right scope. if the vec was created by the scan_notes fn then we would have to return it out of the if let block. this gets more complicated as there will also be outgoing notes and encoded memos that also need returning from the block

Reply:
I share @fluidvanadium 's concern here.   The return of this fn isn't very useful.

This C-style pattern of passing in mutable references and then messing with the state outside the fn doesn't leverage the type system to constrain what can happen.